contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -320,6 +352,589 @@ babelfish_concat_wrapper(PG_FUNCTION_ARGS)
+		 * you must use the date 13/1/97.
+		 */
+		res = (day + 2*month + ((3*(month+1))/(5)) + year +
+					year/4 - year/100 + year/400 + 2) % 7;
+		
+		/* Adjusting the dow accourding to the datefirst guc */
+		return ((res) + 7 - pltsql_datefirst)%7 == 0 ?
+					7 : ((res) + 7 - pltsql_datefirst)%7;
+
+	}
+	else if (strcasecmp(field , "tsql_week") == 0)		/* week number of the year  */
+	{
+		/* returns number of days since 1/1/1970 to 1/1/tm_year */
+		first_day = date2j(tm->tm_year, 1, 1) - UNIX_EPOCH_JDATE;
+
+		/* convert this first day of tm_year to timestamp into tsql_first_day */
+		tsql_first_day = (Timestamp) (first_day - (POSTGRES_EPOCH_JDATE - UNIX_EPOCH_JDATE)) * USECS_PER_DAY;
+
+		first_week_end = 8 - datepart_internal("dow", tsql_first_day, 0, false);
+
+		day_of_year = datepart_internal("doy",timestamp,0, false);
+
+		if(day_of_year <= first_week_end)
+		{
+			/* day of year is less than first_week_end means its a first week */
+			return 1;
+		}
+		else
+		{
+			return (2+(day_of_year - first_week_end - 1) / 7);
+		}
+	}
+	else if(strcasecmp(field , "week") == 0)
+	{
+		return date2isoweek(tm->tm_year, tm->tm_mon, tm->tm_mday);
+	}
+	else if(strcasecmp(field , "millisecond") == 0)
+	{
+		return (fsec1) / 1000;
+	}
+	else if(strcasecmp(field , "microsecond") == 0)
+	{
+		return (fsec1);
+	}
+	else if(strcasecmp(field , "nanosecond") == 0)
+	{
+		return (fsec1) * 1000;
+	}
+	else if(strcasecmp(field , "tzoffset") == 0)
+	{
+		if(general_integer_datatype)
+		{
+			ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				errmsg("The datepart tzoffset is not supported by date function datepart for data type datetime.")));
+			
+			return -1;
+		}
+		else
+		{
+			return (int)df_tz;
+		}
+	}
+	else
+	{
+		ereport(ERROR,
+			(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+			errmsg("\'%s\' is not a recognized datepart option",field)));
+			
+		return -1;
+	}
+	
+	return 1;
+}
+
+
+/*
+ * datepart_internal_datetimeoffset takes datetimeoffset and converts it to
+ * timestamp and calls datepart_internal 
+ */
+
+Datum
+datepart_internal_datetimeoffset(PG_FUNCTION_ARGS)
+{
+	char		*field = text_to_cstring(PG_GETARG_TEXT_PP(0));
+	Timestamp	timestamp;
+	int			df_tz = PG_GETARG_INT32(2);
+
+	timestamp = (Timestamp)(DirectFunctionCall1(common_utility_plugin_ptr->datetimeoffset_timestamp,
+							PG_GETARG_DATUM(1)));
+	
+	timestamp = timestamp + (Timestamp) df_tz * SECS_PER_MINUTE * USECS_PER_SEC;
+	
+	PG_RETURN_INT32(datepart_internal(field, timestamp, (float8)df_tz, false));
+}
+
+/*
+ * datepart_internal_date takes date and converts it to
+ * timestamp and calls datepart_internal 
+ */
+
+Datum
+datepart_internal_date(PG_FUNCTION_ARGS)
+{
+	char		*field = text_to_cstring(PG_GETARG_TEXT_PP(0));
+	Timestamp	timestamp;
+	int			df_tz = PG_GETARG_INT32(2);
+	DateADT		date_arg;
+	date_arg = PG_GETARG_DATEADT(1);
+	timestamp = DirectFunctionCall1(date_timestamp, date_arg);
```

Comment:
DirectFunctionCall1 expects datum argument so we can directly pass PG_GETARG(1) to DirectFunctionCall1